File: packages/protocol/src/subdomain.ts
import { customAlphabet } from "nanoid";
import { PROTOCOL } from "./constants.js";

export const generateSubdomainId = customAlphabet(
  PROTOCOL.SUBDOMAIN_ALPHABET,
  PROTOCOL.SUBDOMAIN_LENGTH,
);

export const generateRequestId = customAlphabet(
  PROTOCOL.SUBDOMAIN_ALPHABET,
  PROTOCOL.REQUEST_ID_LENGTH,
);

const generateSuffix = customAlphabet(
  PROTOCOL.SUBDOMAIN_ALPHABET,
  PROTOCOL.SUBDOMAIN_SUFFIX_LENGTH,
);

/**
 * Build a subdomain from a user-provided prefix by sanitizing it and
 * appending a random suffix: `my-app` → `my-app-x7k2m4`.
 * Falls back to a fully random ID if the prefix is empty after sanitizing.
 */
export function buildCustomSubdomain(prefix: string): string {
  const sanitized = prefix
    .toLowerCase()
    .replace(/[^a-z0-9-]/g, "")
    .replace(/^-+|-+$/g, "");
  if (!sanitized) return generateSubdomainId();
  return `${sanitized}-${generateSuffix()}`;
}

const SUBDOMAIN_PATTERN = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/;
const MAX_SUBDOMAIN_LENGTH = 63; // DNS label limit

export function validateSubdomain(
  subdomain: string,
): { ok: true } | { ok: false; reason: string } {
  if (subdomain.length < 1 || subdomain.length > MAX_SUBDOMAIN_LENGTH) {
    return { ok: false, reason: "Subdomain must be 1-63 characters" };
  }
  if (!SUBDOMAIN_PATTERN.test(subdomain)) {
    return {
      ok: false,
      reason:
        "Subdomain must contain only lowercase letters, digits, and hyphens",
    };
  }
  return { ok: true };
}

File: apps/cli-go/internal/protocol/subdomain.go
package protocol

import (
	"crypto/rand"
	"regexp"
	"strings"
)

var (
	validSubdomainRe = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]*[a-z0-9])?$`)
	sanitizeRe       = regexp.MustCompile(`[^a-z0-9-]`)
)

// generateRandomString produces a cryptographically random string of the given
// length using characters from SubdomainAlphabet.
func generateRandomString(length int) string {
	buf := make([]byte, length)
	if _, err := rand.Read(buf); err != nil {
		panic("crypto/rand: " + err.Error())
	}
	alphabetLen := len(SubdomainAlphabet)
	result := make([]byte, length)
	for i, b := range buf {
		result[i] = SubdomainAlphabet[int(b)%alphabetLen]
	}
	return string(result)
}

// GenerateSubdomainID returns a random subdomain identifier of SubdomainLength characters.
func GenerateSubdomainID() string {
	return generateRandomString(SubdomainLength)
}

// GenerateRequestID returns a random request identifier of RequestIDLength characters.
func GenerateRequestID() string {
	return generateRandomString(RequestIDLength)
}

// BuildCustomSubdomain takes a user-supplied prefix, sanitizes it, and appends
// a random suffix. If the prefix is empty or contains only invalid characters,
// a fully random subdomain is returned instead.
func BuildCustomSubdomain(prefix string) string {
	cleaned := sanitizeRe.ReplaceAllString(strings.ToLower(prefix), "")
	cleaned = strings.Trim(cleaned, "-")

	if cleaned == "" {
		return GenerateSubdomainID()
	}

	suffix := generateRandomString(SubdomainSuffixLength)
	return cleaned + "-" + suffix
}

// ValidateSubdomain checks whether a subdomain string is valid.
// It returns (true, "") on success, or (false, reason) on failure.
func ValidateSubdomain(subdomain string) (bool, string) {
	if len(subdomain) == 0 {
		return false, "subdomain must not be empty"
	}
	if len(subdomain) > 63 {
		return false, "subdomain must be at most 63 characters"
	}
	if !validSubdomainRe.MatchString(subdomain) {
		return false, "subdomain must contain only lowercase alphanumeric characters and hyphens, and must not start or end with a hyphen"
	}
	return true, ""
}
